server/http_test.go

```diff
@@ -332,6 +332,125 @@ func TestHandleAuthFuncResponsesMultipleRedirectURLs(t *testing.T) {
 	}
 }
 
+func TestHandleTokenFunc(t *testing.T) {
+
+	fx, err := makeTestFixtures()
+	if err != nil {
+		t.Errorf("could not run test fixtures")
+	}
+
+	tests := []struct {
+		query    url.Values
+		creds    string
+		wantCode int
+	}{
+		// bad grant_type
+		{
+			query: url.Values{
+				"grant_type": []string{"invalid!"},
+				"state":      []string{""},
+				"code":       []string{"someCode"},
+			},
+			creds:    "XXX:" + base64.URLEncoding.EncodeToString([]byte("secrete")),
+			wantCode: http.StatusBadRequest,
+		},
+
+		// authorization_code needs code param
+		{
+			query: url.Values{
+				"grant_type": []string{"authorization_code"},
+				"state":      []string{""},
+			},
+			creds:    "XXX:" + base64.URLEncoding.EncodeToString([]byte("secrete")),
+			wantCode: http.StatusBadRequest,
+		},
+
+		// empty code
+		{
+			query: url.Values{
+				"grant_type": []string{"authorization_code"},
+				"state":      []string{""},
+				"code":       []string{""},
+			},
+			creds:    "XXX:" + base64.URLEncoding.EncodeToString([]byte("secrete")),
+			wantCode: http.StatusBadRequest,
+		},
+
+		// valid code but bad creds
+		{
+			query: url.Values{
+				"grant_type": []string{"authorization_code"},
+				"state":      []string{""},
+				"code":       []string{"code-2"},
+			},
+			creds:    "XASD:" + base64.URLEncoding.EncodeToString([]byte("failsecrete")),
+			wantCode: http.StatusUnauthorized,
+		},
+
+		// bad code
+		{
+			query: url.Values{
+				"grant_type": []string{"authorization_code"},
+				"state":      []string{""},
+				"code":       []string{"asdasd"},
+			},
+			creds:    "XXX:" + base64.URLEncoding.EncodeToString([]byte("secrete")),
+			wantCode: http.StatusBadRequest,
+		},
+
+		// OK testcase
+		{
+			query: url.Values{
+				"grant_type": []string{"authorization_code"},
+				"state":      []string{""},
+				"code":       []string{"code-2"},
+			},
+			creds:    "XXX:" + base64.URLEncoding.EncodeToString([]byte("secrete")),
+			wantCode: http.StatusOK,
+		},
+	}
+
+	for i, tt := range tests {
+		hdlr := handleTokenFunc(fx.srv)
+		w := httptest.NewRecorder()
+
+		req, err := http.NewRequest("POST", "http://example.com/token", strings.NewReader(tt.query.Encode()))
+		if err != nil {
+			t.Errorf("unable to create HTTP request, error=%v", err)
```

Comment:
Need a `continue` statement here so this skips to the next test case on an error. If not `req` will be nil and this test will produce panics.